openvdb/points/PointScatter.h

```diff
@@ -0,0 +1,676 @@
+                    if (value == 0) leaf.setValueOff(iter.pos());
+                    else            offset += value;
+                }
+                // @note can't use iter.setValue(offset) on point grids
+                leaf.setOffsetOnly(iter.pos(), offset);
+            }
+
+            if (offset != 0) BaseT::generatePositions(leaf, offset);
+        }, /*threaded*/false);
+
+        if (mPointsPerVoxel == 0) tools::pruneInactive(BaseT::mPoints->tree());
+        BaseT::end();
+    }
+
+    /// @brief Print information about the scattered points
+    /// @parm name  A name to insert into the printed info
+    /// @parm os    The output stream
+    void print(const std::string &name, std::ostream& os = std::cout) const
+    {
+        if (!BaseT::mPoints) return;
+        const Index64 points = pointCount(BaseT::mPoints->tree());
+        const Index64 voxels = BaseT::mPoints->activeVoxelCount();
+
+        os << "Uniformly scattered " << points << " points into " << voxels
+           << " active voxels in \"" << name << "\" corresponding to "
+           << (double(points) / double(voxels)) << " points per voxel." << std::endl;
+    }
+
+private:
+    const Index64 mPointCount;
+    Index32 mPointsPerVoxel;
+};
+
+
+template <typename RandGenT,
+          typename ValueType,
+          typename CodecT,
+          typename PointDataGridT,
+          typename InterruptT>
+class BasePointScatter
+{
+public:
+    using RandomGenerator = math::Rand01<ValueType, RandGenT>;
+    using TreeT = typename PointDataGridT::TreeType;
+    using LeafNodeT = typename TreeT::LeafNodeType;
+    using LeafManagerT = tree::LeafManager<TreeT>;
+
+    /// @brief Retrieve the scattered points
+    /// @note will contain a nullptr if the derived operator
+    /// has not been called
+    inline typename PointDataGridT::Ptr points()
+    {
+        return mPoints;
+    }
+
+    /// @brief initialise the RNG seed
+    /// @parm seed   The seed for the random number generator
+    inline void setSeed(const unsigned int seed)
+    {
+        mSeed = seed;
+        mRand01.setSeed(mSeed);
+    }
+
+    /// @brief set the spread value
+    /// @parm spread   The spread of points from the voxel center from 0 to 1
+    inline void setSpread(const float spread)
+    {
+        assert(spread >= 0.0f && spread <= 1.0f);
+        mSpread = spread;
+    }
+
+    /// @brief Return a random floating point number between zero and one
+    inline double getRand01() const
+    {
+        return mRand01();
+    }
+
+    /// @brief check to see if an interrupt has been requested
+    inline bool wasInterrupted()
+    {
+        return util::wasInterrupted<InterruptT>(mInterrupter);
+    }
+
+    inline void generatePositions(LeafNodeT& leaf, const Index64& count) const
+    {
+        leaf.initializeAttributes(mAttributeDescriptor, count);
+
+        // directly expand to avoid needlessly setting uniform values in the
+        // write handle
```

Comment:
Highly precise optimisation.

Reply:
This was a significant performance increase for large numbers of points - it seems like a reasonable thing to do and it's nice the API allows it

Reply:
That's surprising it made that big a difference, but at least we added it for a reason. :)

(Incidentally, think there are a couple of minor IndexFilter/IndexIterator performance tweaks which might not have been submitted yet, need to double check that).